### ISSUE TITLE
Can't connect with self signed https certificate

### ISSUE THREAD

**Issue by gguzzina** [GrakovNe/lissen-android#29]
I am running Audiobookshelf inside a Yunohost installation, and that means that I need to connect to Audiobookshelf ober https using a self-signed certificate.

I don't have a public IP, so I can't use Let's Encrypt.

I have imported the CA certificate of my Yunohost installation inside my Android phone's "trusted credentials".

Stock Audiobookshelf app works.

**Comment by GrakovNe** [comment]
Hi!

I see the issue. It's really needs to be implemented 

Could you give me an access for your library with CA certificate for the debug please? You may create a test credentials with no books just to give me a log into server

Thanks!

**Comment by gguzzina** [comment]
Well, as I don't have a public IP (ISP oddities) you'd have to come to my home. Sorry 😃

**Comment by GrakovNe** [comment]
no worries, i'll try with my own self-signed certificate soon!

**Comment by pratiksumbe** [comment]
/what is server ORL & where will I find it?

**Comment by pratiksumbe** [comment]
> /what is server ORL & where will I find it?

URL

**Comment by GrakovNe** [comment]
@pratiksumbe you have to host your own ABS server to get access in general. Audiobookshelf is a self-hosting project where anyone able to host own instance with own data and use it for own requirements

You may read how to by the official ABS doc: https://www.audiobookshelf.org/docs

also, for demo requirements you may connect to https://demo.lissenapp.org with demo/demo credentials via Lissen or any other client (including the official one)

**Comment by GrakovNe** [comment]
Finally implemented in [release 1.1.1](https://github.com/GrakovNe/lissen-android/releases/tag/1.1.1).

To replicate an environment where the ABS server uses a self-signed SSL certificate, I did the following:

1. I created a self-signed certificate and a self-signed Certificate Authority (CA)

```
openssl genrsa -out myCA.key 2048
openssl req -x509 -new -nodes -key myCA.key -sha256 -days 1024 -out myCA.pem -subj "/C=RU/ST=State/L=City/O=Organization/OU=OrgUnit/CN=MyRootCA"
openssl genrsa -out server.key 2048
openssl req -new -key server.key -out server.csr -subj "/C=RU/ST=State/L=City/O=Organization/OU=OrgUnit/CN=192.168.1.12"
```

where 192.168.1.12 is the host of my brand-new ABS server installation

2. I created the following configuration file:

```
authorityKeyIdentifier=keyid,issuer
basicConstraints=CA:FALSE
keyUsage = digitalSignature, keyEncipherment
extendedKeyUsage = serverAuth
subjectAltName = @alt_names

[alt_names]
IP.1 = 192.168.1.12
```

and signed the certificate using the provided configuration

```
openssl x509 -req -in server.csr -CA myCA.pem -CAkey myCA.key -CAcreateserial -out server.crt -days 500 -sha256 -extfile server.ext
```

3. Then I installed myCA.pem as a trusted credential

![2024-11-12 22 29 21](https://github.com/user-attachments/assets/e83558ec-2fc3-4541-b70e-37a2b2cbc0c4)

5. I used the following proxy for ABS to 'secure' it with a certificate signed by the same CA

```
socat OPENSSL-LISTEN:13375,cert=server.pem,key=server.key,verify=0,fork TCP:localhost:13378
```

This configuration made it impossible for the application to connect to the server with a self-signed SSL certificate, despite the root certificate being present in the system on the client device.

However, this allowed me to develop and debug a fix enabling the application to use the system certificate store to establish a connection to servers.

The fix itself is very simple

```
fun OkHttpClient.Builder.withTrustedCertificates(): OkHttpClient.Builder {
    return try {
        val trustManager = getSystemTrustManager()
        val sslContext = getSystemSSLContext(trustManager)
        this.sslSocketFactory(sslContext.socketFactory, trustManager)

        this
    } catch (ex: Exception) {
        this
    }
}
```

and then

```
...
OkHttpClient
        .Builder()
        .withTrustedCertificates()
        .addInterceptor(
            HttpLoggingInterceptor().apply {
                level = HttpLoggingInterceptor.Level.NONE
            }
        )
...
```

Now you can connect to the server even with a self-signed certificate.

Let me show you

![2024-11-12 22 34 51](https://github.com/user-attachments/assets/461f1098-194f-4ea3-8cef-7e7bc5b01906)
![2024-11-12 22 34 47](https://github.com/user-attachments/assets/0d38ada0-2ded-452e-b9bf-47149a9b29a9)
![2024-11-12 22 34 54](https://github.com/user-attachments/assets/2ce60549-19b0-4801-a75a-ba8443a4b20d)

It works for:

Connecting to the ABS server and reading server metadata
Playing streams from the ABS server while connected
However, it doesn't work for caching books locally to play without server access.

This is because the initial implementation for downloading, from the first app version, uses the [Android DownloadManager](https://developer.android.com/reference/android/app/DownloadManager), which doesn’t support self-signed certificates.

The issue for redesigning the download feature is tracked here: [#50](https://github.com/GrakovNe/lissen-android/issues/50).

The new implementation will avoid using DownloadManager, as it lacks the flexibility needed for the app’s requirements.

I'd be glad if you could check whether the app works with your own SSL certificate!
